### PR TITLE
[Core][Geometries] Missing or not well defined DomainSize functions

### PR DESCRIPTION
**Pull request by loumalouomega**
To avoid overload of changes #3729

### CODE REVIEW

File: kratos/geometries/geometry.h
Comment (loumalouomega) [inline kratos/geometries/geometry.h]: Additionally to what is defined in #3729 I added this generic version of DomainSIze

**Comment by loumalouomega** [comment]
@KratosMultiphysics/implementation-committee

**Comment by loumalouomega** [comment]
Anyone?

File: kratos/geometries/quadrilateral_3d_9.h
Comment (philbucher) [inline kratos/geometries/quadrilateral_3d_9.h]: ?

File: kratos/geometries/quadrilateral_3d_9.h
Comment (philbucher) [inline kratos/geometries/quadrilateral_3d_9.h]: Those comments are not really reassuring this works...
Are there tests?

File: kratos/geometries/quadrilateral_3d_9.h
Comment (philbucher) [inline kratos/geometries/quadrilateral_3d_9.h]: ?

File: kratos/geometries/quadrilateral_3d_8.h
Comment (philbucher) [inline kratos/geometries/quadrilateral_3d_8.h]: ?

**Review by philbucher — COMMENTED** [review]
many questionable comments

also, did you run ALL tests in the main-apps?

File: kratos/geometries/quadrilateral_3d_9.h
Comment (loumalouomega) [reply kratos/geometries/quadrilateral_3d_9.h]: This is a old comment put it in correct Doxygen format

File: kratos/geometries/quadrilateral_3d_9.h
Comment (loumalouomega) [reply kratos/geometries/quadrilateral_3d_9.h]: These comments are old comments in Doxygen format

File: kratos/geometries/quadrilateral_3d_9.h
Comment (loumalouomega) [reply kratos/geometries/quadrilateral_3d_9.h]: Again the old comments in Doxygen

File: kratos/geometries/quadrilateral_3d_8.h
Comment (loumalouomega) [reply kratos/geometries/quadrilateral_3d_8.h]: Again, updating to correct Doxygen

**Comment by loumalouomega** [comment]
Yes @philbucher I run all the tests of the core applications (fluid, structural, conv-diff, meshing)

**Review by GuillermoCasas — COMMENTED** [review]
Is it really necessary to have Volume() defined for flat elements? Would it not be preferable to give an error?

**Comment by loumalouomega** [comment]
@GuillermoCasas I think this is very "controversial discussion". We should define a consistent way to define this (and update the current use to avoid problems). 

A first step would be to add a warning instead of an error, and tell the developer to switch to DomainSize() (I was using Volume as general method instead of DomainSize until recently). That way the developers can adapt the code without extra problems. 

The second one should be the definition of the values to be given, for example many 3d elements returns the volume as Area method, when should be returning the sum of the faces areas (in my opinion).

The third step should be removing the duplicated code of the volumes and areas computed *a la* FEM. Except some very simple geometries all geometries compute the Volume and Area using the FEM integration. I would remove the methods in the derived geometries and adding it to the  base one (instead of the current error) and just add the methods in case that the Area or Volume can be compute analytically

Involving @KratosMultiphysics/technical-committee @KratosMultiphysics/implementation-committee @KratosMultiphysics/design-committee

**Comment by loumalouomega** [comment]
> @GuillermoCasas I think this is very "controversial discussion". We should define a consistent way to define this (and update the current use to avoid problems).
> 
> A first step would be to add a warning instead of an error, and tell the developer to switch to DomainSize() (I was using Volume as general method instead of DomainSize until recently). That way the developers can adapt the code without extra problems.
> 
> The second one should be the definition of the values to be given, for example many 3d elements returns the volume as Area method, when should be returning the sum of the faces areas (in my opinion).
> 
> The third step should be removing the duplicated code of the volumes and areas computed _a la_ FEM. Except some very simple geometries all geometries compute the Volume and Area using the FEM integration. I would remove the methods in the derived geometries and adding it to the base one (instead of the current error) and just add the methods in case that the Area or Volume can be compute analytically
> 
> Involving @KratosMultiphysics/technical-committee @KratosMultiphysics/implementation-committee @KratosMultiphysics/design-committee

Moving this discussion to the future, can we check it this?, it is necessary for #3729, and @RiccardoRossi suggested to merge ASAP #3729

File: kratos/geometries/geometry.h
Comment (pooyan-dadvand) [reply kratos/geometries/geometry.h]: The local space dimension for a 2D triangle would be 3 and gives a wrong result. I would use `Dimension()` to be more precise

File: kratos/geometries/quadrilateral_2d_4.h
Comment (pooyan-dadvand) [inline kratos/geometries/quadrilateral_2d_4.h:478]: Are you sure? The volume should be zero. Isn't it?

File: kratos/geometries/quadrilateral_2d_8.h
Comment (pooyan-dadvand) [inline kratos/geometries/quadrilateral_2d_8.h]: ?

File: kratos/geometries/quadrilateral_2d_8.h
Comment (pooyan-dadvand) [inline kratos/geometries/quadrilateral_2d_8.h]: again

File: kratos/geometries/quadrilateral_2d_8.h
Comment (pooyan-dadvand) [inline kratos/geometries/quadrilateral_2d_8.h]: Do we need this comment? I suppose you are modifying it to something more suitable right now

File: kratos/geometries/quadrilateral_2d_9.h
Comment (pooyan-dadvand) [inline kratos/geometries/quadrilateral_2d_9.h]: ?

File: kratos/geometries/quadrilateral_2d_9.h
Comment (pooyan-dadvand) [inline kratos/geometries/quadrilateral_2d_9.h:452]: again

File: kratos/geometries/quadrilateral_3d_9.h
Comment (pooyan-dadvand) [reply kratos/geometries/quadrilateral_3d_9.h]: I would update the comments if they are not correct anymore

**Review by pooyan-dadvand — CHANGES_REQUESTED** [review]
What is the rational behind returning the area as volume in 2D shapes?

File: kratos/geometries/quadrilateral_2d_4.h
Comment (loumalouomega) [reply kratos/geometries/quadrilateral_2d_4.h:478]: Yes

**Comment by pooyan-dadvand** [comment]
> Moving this discussion to the future, can we check it this?, it is necessary for #3729, and @RiccardoRossi suggested to merge ASAP #3729 

I don't agree if the volume should return area and we cannot accelarate some PRs because other PR is using it. you may separate the dependencies if you want to unblock the other one.

**Comment by loumalouomega** [comment]
Maybe I explained me wrong. What I was saying is that if we mix discussions
it can be more difficult to reach a consensus.

Related to separate PRs, this PR has as objective that. To separate the
update of the domain size and the lumped factors.

When I said ASAP I was referring to what @RiccardoRossi said.

**Comment by pooyan-dadvand** [comment]
I see, thanks for clarification!

**Comment by loumalouomega** [comment]
Ping, @pooyan-dadvand, there is something missing in this PR?

File: kratos/geometries/quadrilateral_3d_4.h
Comment (pooyan-dadvand) [inline kratos/geometries/quadrilateral_3d_4.h]: Why committing a comment?

File: kratos/geometries/quadrilateral_3d_4.h
Comment (pooyan-dadvand) [inline kratos/geometries/quadrilateral_3d_4.h:456]: Why you are changing the implementation? It seems to be slower than previous one.

**Review by pooyan-dadvand — CHANGES_REQUESTED** [review]
One general comment: Why should we change the behaviour of Volume for 2D geometries

File: kratos/geometries/quadrilateral_3d_4.h
Comment (loumalouomega) [reply kratos/geometries/quadrilateral_3d_4.h]: This PR was originally part of #3729, so maybe many things were mixed

File: kratos/geometries/quadrilateral_3d_4.h
Comment (loumalouomega) [reply kratos/geometries/quadrilateral_3d_4.h:456]: The previous one wasn't working in general, for warped quadrilateral 3d 4

File: kratos/geometries/quadrilateral_3d_4.h
Comment (loumalouomega) [reply kratos/geometries/quadrilateral_3d_4.h]: I will check it again

**Comment by msandre** [comment]
> Moving this discussion to the future, can we check it this?, it is necessary for #3729, and @RiccardoRossi suggested to merge ASAP #3729

I understood this to mean that you are not needing the @KratosMultiphysics/design-committee currently. If this is wrong please correct.

Besides that I like the changes. Thanks :+1:

File: kratos/tests/cpp_tests/geometries/test_quadrilateral_3d_4.cpp
Comment (RiccardoRossi) [inline kratos/tests/cpp_tests/geometries/test_quadrilateral_3d_4.cpp:178]: which one is the "correct" solution? 

We should test this with very many decimals ...

Comment (loumalouomega) [reply kratos/tests/cpp_tests/geometries/test_quadrilateral_3d_4.cpp:178]: I did this PR long time ago, let me check in any case

Comment (GuillermoCasas) [reply kratos/tests/cpp_tests/geometries/test_quadrilateral_3d_4.cpp:178]: Related to #4305.

File: kratos/geometries/quadrilateral_3d_4.h
Comment (RiccardoRossi) [reply kratos/geometries/quadrilateral_3d_4.h:456]: +1

File: kratos/tests/cpp_tests/geometries/test_quadrilateral_3d_4.cpp
Comment (RiccardoRossi) [reply kratos/tests/cpp_tests/geometries/test_quadrilateral_3d_4.cpp:178]: the point is that this should have very many decimals (around 15) since it is around 1.0

File: kratos/tests/cpp_tests/geometries/test_triangle_2d_6.cpp
Comment (RiccardoRossi) [inline kratos/tests/cpp_tests/geometries/test_triangle_2d_6.cpp]: what is the value of tolerance here? it should be very small

Comment (loumalouomega) [reply kratos/tests/cpp_tests/geometries/test_triangle_2d_6.cpp]: It is defined on the beginning of the file, 1.0e-8

Comment (RiccardoRossi) [reply kratos/tests/cpp_tests/geometries/test_triangle_2d_6.cpp]: i would make this much stricter.

Comment (loumalouomega) [reply kratos/tests/cpp_tests/geometries/test_triangle_2d_6.cpp]: It would affect other tests

**Comment by RiccardoRossi** [comment]
i am not proposing to change the value TOLERANCE

i am telling that depending on the test you may be much stricter in the tolerance you require.
in principle simple tests should be much closer to machine precision than things that require a lot of calculations

**Comment by loumalouomega** [comment]
> i am not proposing to change the value TOLERANCE
> 
> i am telling that depending on the test you may be much stricter in the tolerance you require.
> in principle simple tests should be much closer to machine precision than things that require a lot of calculations

Done @RiccardoRossi

**Comment by msandre** [comment]
> Is it really necessary to have Volume() defined for flat elements? Would it not be preferable to give an error?

Similar to the question, does it make sense to have Normal() defined for volume elements? It's definitely a code smell which suggests some refactoring of the geometry is needed. It would be great if we could address this together with the Nurbs discussion in #4724 which appears to be stretching the geometry interface even further.

**Comment by pooyan-dadvand** [comment]
I agree that geometry needs some refactoring and I agree that the introduction of Nurbs is a good moment to do it. :+1:

**Comment by philbucher** [comment]
What is the status of this?

**Comment by loumalouomega** [comment]
> What is the status of this?

Some refactor is needed

**Comment by loumalouomega** [comment]
> > What is the status of this?
> 
> Some refactor is needed

I don't know now the current status of this (I solved conflicts)

File: kratos/geometries/quadrilateral_2d_4.h
Comment (rubenzorrilla) [reply kratos/geometries/quadrilateral_2d_4.h:478]: As it is commented by @loumalouomega in the discussion, I'd throw a warning in here. This would reveal the misuses of the `Volume` function to ease the update to `DomainSize`.

File: kratos/geometries/geometry.h
Comment (rubenzorrilla) [inline kratos/geometries/geometry.h]: We should agree about this @KratosMultiphysics/technical-committee. IMO, it makes sense.

File: kratos/geometries/quadrilateral_3d_4.h
Comment (rubenzorrilla) [inline kratos/geometries/quadrilateral_3d_4.h]: Same comment, I'd throw a warning in here.

File: kratos/geometries/triangle_2d_6.h
Comment (rubenzorrilla) [inline kratos/geometries/triangle_2d_6.h]: Same warning comment.

File: kratos/geometries/triangle_3d_3.h
Comment (rubenzorrilla) [inline kratos/geometries/triangle_3d_3.h]: Same warning comment.

File: kratos/geometries/triangle_3d_6.h
Comment (rubenzorrilla) [inline kratos/geometries/triangle_3d_6.h]: Same warning comment.

File: kratos/geometries/triangle_3d_6.h
Comment (rubenzorrilla) [inline kratos/geometries/triangle_3d_6.h:467]: Check this @KratosMultiphysics/technical-committee. IMO, it makes sense, previous implementation was assuming that the nodes in the mid of the edges were aligned with the vertex nodes. This is more general.

File: kratos/tests/cpp_tests/geometries/test_geometry.h
Comment (rubenzorrilla) [inline kratos/tests/cpp_tests/geometries/test_geometry.h:40]: Wouldn't this give us machine precision issues? @pooyan-dadvand @roigcarlo

File: kratos/tests/cpp_tests/geometries/test_quadrilateral_3d_4.cpp
Comment (rubenzorrilla) [reply kratos/tests/cpp_tests/geometries/test_quadrilateral_3d_4.cpp:178]: I agree with @RiccardoRossi 's observation. The tolerance in this case must be smaller.

**Review by rubenzorrilla — COMMENTED** [review]
@KratosMultiphysics/technical-committee I put a comment on each line we should have a look to. The rest of changes are either documentation or aesthetic. I should also say that the proposed changes make sense for me.

File: kratos/geometries/triangle_3d_6.h
Comment (loumalouomega) [reply kratos/geometries/triangle_3d_6.h]: Done (idem from the previous ones)

File: kratos/tests/cpp_tests/geometries/test_geometry.h
Comment (loumalouomega) [reply kratos/tests/cpp_tests/geometries/test_geometry.h:40]: Do you mean 32bit vs 64 bit?¿

File: kratos/geometries/quadrilateral_2d_4.h
Comment (pooyan-dadvand) [reply kratos/geometries/quadrilateral_2d_4.h:478]: @KratosMultiphysics/technical-committee suggests to add a debug warning to all these methods with trivial results which should not be called in normal situation.

File: kratos/geometries/quadrilateral_3d_4.h
Comment (pooyan-dadvand) [reply kratos/geometries/quadrilateral_3d_4.h]: same answer!

File: kratos/geometries/triangle_3d_6.h
Comment (pooyan-dadvand) [reply kratos/geometries/triangle_3d_6.h:467]: @KratosMultiphysics/technical-committee agrees with this change but suggests to change the temp to `array_1d`

File: kratos/tests/cpp_tests/geometries/test_geometry.h
Comment (pooyan-dadvand) [reply kratos/tests/cpp_tests/geometries/test_geometry.h:40]: I am not sure in 32bit but in 64bit it is ok

File: kratos/tests/cpp_tests/geometries/test_quadrilateral_3d_4.cpp
Comment (pooyan-dadvand) [reply kratos/tests/cpp_tests/geometries/test_quadrilateral_3d_4.cpp:178]: why the tolerance and results are changed?

**Comment by rubenzorrilla** [comment]
@loumalouomega I just accidentally came across with this. Which is the status? At first glance, it seems to me that some of the changes might still have sense.

**Comment by loumalouomega** [comment]
> @loumalouomega I just accidentally came across with this. Which is the status? At first glance, it seems to me that some of the changes might still have sense.

I didn't remember about this. I will take a look

File: kratos/geometries/quadrilateral_interface_3d_4.h
Comment (loumalouomega) [inline kratos/geometries/quadrilateral_interface_3d_4.h]:
```suggestion
        return 0.0;
```

**Comment by loumalouomega** [comment]
> @loumalouomega I just accidentally came across with this. Which is the status? At first glance, it seems to me that some of the changes might still have sense.

Looks like is mergeable, what is the current problem?

**Comment by loumalouomega** [comment]
Look @AlejandroCornejo this PR is 4 years old, I have the record

**Comment by rubenzorrilla** [comment]
> > @loumalouomega I just accidentally came across with this. Which is the status? At first glance, it seems to me that some of the changes might still have sense.
> 
> Looks like is mergeable, what is the current problem?

Well, there are some behavior changes that should be reconsidered IMO.

**Comment by loumalouomega** [comment]
> > > @loumalouomega I just accidentally came across with this. Which is the status? At first glance, it seems to me that some of the changes might still have sense.
> > 
> > 
> > Looks like is mergeable, what is the current problem?
> 
> Well, there are some behavior changes that should be reconsidered IMO.

Which ones? (PR is too old, is very difficult to see old comments)

File: kratos/tests/cpp_tests/geometries/test_triangle_3d_6.cpp
Comment (rubenzorrilla) [inline kratos/tests/cpp_tests/geometries/test_triangle_3d_6.cpp:64]: This one for instance.

Comment (loumalouomega) [reply kratos/tests/cpp_tests/geometries/test_triangle_3d_6.cpp:64]: Yes, but if the origin of the problem is inconsistency then we will have always a behaviour change because by definition an inconsistency implies that something is not behaving as is supposed

Comment (rubenzorrilla) [reply kratos/tests/cpp_tests/geometries/test_triangle_3d_6.cpp:64]: I'm not discussing if this whether makes sense or not, my point is that, in my opinion, it is worth discussing these again as the PR is quite old and these are crucial classes. @KratosMultiphysics/technical-committee

Comment (loumalouomega) [reply kratos/tests/cpp_tests/geometries/test_triangle_3d_6.cpp:64]: Understandable, just explaining my point

File: kratos/geometries/geometry.h
Comment (pooyan-dadvand) [inline kratos/geometries/geometry.h]: This should be LocalSpaceDimension()

**Comment by pooyan-dadvand** [comment]
After a new discussion about it at @KratosMultiphysics/technical-committee agrees with the changes but should be done in two separate steps in two separate PRs:
1. Giving the warning WITHOUT changing the current behavior.  The warning will say that it will be changed in 6 months
2. After the period of the change, in a new PR changing the behavior and remove the warning and put an error

**Comment by loumalouomega** [comment]
Done @pooyan-dadvand and @KratosMultiphysics/technical-committee

**Comment by pooyan-dadvand** [comment]
@loumalouomega Please note that it will be replaced with Error and not with zero!

**Comment by loumalouomega** [comment]
> @loumalouomega Please note that it will be replaced with Error and not with zero!

Okay

**Comment by loumalouomega** [comment]
This is ready @pooyan-dadvand

**Comment by pooyan-dadvand** [comment]
I just noticed that in the warning we are encouraging to use DomainSize. Is it optimum or shall we put Area() or other respective method (in function of geometry type) as alternative?

File: kratos/geometries/geometry.h
Comment (loumalouomega) [inline kratos/geometries/geometry.h:1385]: Here the check is done @pooyan-dadvand

**Comment by loumalouomega** [comment]
This is ready @pooyan-dadvand

File: kratos/geometries/sphere_3d_1.h
Comment (rubenzorrilla) [inline kratos/geometries/sphere_3d_1.h:404]: It was already here so I'd leave it as it is, but I think that in this case `DomainSize` does make sense.

File: kratos/geometries/geometry.h
Comment (rubenzorrilla) [inline kratos/geometries/geometry.h:1383]: Just putting it in here as a reminder for the future, but this method should no longer be `virtual`.

File: kratos/geometries/triangle_3d_6.h
Comment (rubenzorrilla) [reply kratos/geometries/triangle_3d_6.h:467]: > @KratosMultiphysics/technical-committee agrees with this change but suggests to change the temp to `array_1d`

This can't be done as the size is not known at compile time. Resolving.

**Review by rubenzorrilla — APPROVED** [review]
Changes are for good and ready. Thanks also for the documentation addition.

**Comment by loumalouomega** [comment]
@pooyan-dadvand this is ready

File: kratos/geometries/triangle_3d_6.h
Comment (pooyan-dadvand) [reply kratos/geometries/triangle_3d_6.h:467]: We are in triangle 6 and we know the default integration method. So we can hardcode the size. I think for such a low level function, such a performance improvement can be justified.

Comment (loumalouomega) [reply kratos/geometries/triangle_3d_6.h:467]: In that case we can do the same for any other integration. I would do in an independent PR, this PR is already old and it would make even more confusing the changes

**Review by pooyan-dadvand — APPROVED** [review]
I think that we can do the optimization that I am suggesting in a separate PR.

**Pull request merged**
